File: src/annotation.go
// annotation.go
//
// Copyright (c) 2026 PDFjet Software
// Licensed under the MIT License. See LICENSE file in the project root.

package pdfjet

// annotationType constants
const (
	annotationLink           = "Link"
	annotationFileAttachment = "FileAttachment"
	annotationPolygon        = "Polygon"
	annotationCircle         = "Circle"
	annotationSquare         = "Square"
	annotationText           = "Text"
)

// annotationObject represents a PDF annotation object.
type annotationObject struct {
	objNumber      int
	annotationType string
	x1             float32
	y1             float32
	x2             float32
	y2             float32
	vertices       []float32
	fillColor      [3]float32
	opacity        float32
	title          string
	contents       string
	uri            string
	key            string
	language       string
	actualText     string
	altDescription string
	fileAttachment *FileAttachment // Assuming FileAttachment type exists elsewhere
	// Set once the annotation has been written with a /StructParent key.
	structParentWritten bool
}

// setDescriptionFallback fills in the actual text and the alternative
// description of an annotation that was created without them. The other ports
// do this in the annotationObject constructor; annotations here are built as struct
// literals, so Page.AddAnnotation applies it instead.
// A link created from a destination name has no uri to fall back on,
// so use the name itself rather than leaving the link undescribed.
func (annotation *annotationObject) setDescriptionFallback() {
	fallback := annotation.uri
	if fallback == "" {
		fallback = annotation.key
	}
	if annotation.actualText == "" {
		annotation.actualText = fallback
	}
	if annotation.altDescription == "" {
		annotation.altDescription = fallback
	}
}

File: net/pdfjet/Annotation.cs
/**
 * Annotation.cs
 *
 * Copyright (c) 2026 PDFjet Software
 * Licensed under the MIT License. See LICENSE file in the project root.
 */
using System;

/**
 * Used to create PDF annotation objects.
 */
namespace PDFjet.NET {
internal class Annotation {
    public static readonly String Link = "Link";
    public static readonly String FileAttachment = "FileAttachment";
    public static readonly String Polygon = "Polygon";
    public static readonly String Circle = "Circle";
    public static readonly String Square = "Square";
    public static readonly String Text = "Text";

    internal int objNumber;
    internal String annotationType = null;
    internal float x1 = 0f;
    internal float y1 = 0f;
    internal float x2 = 0f;
    internal float y2 = 0f;
    internal float[] vertices = null;
    internal float[] fillColor = null;
    internal float transparency = 0f;
    internal String title = null;
    internal String contents = null;
    internal String uri = null;
    internal String key = null;
    internal String language = null;
    internal String actualText = null;
    internal String altDescription = null;
    internal FileAttachment fileAttachment = null;

    /**
     * This class is used to create annotation objects.
     *
     * @param annotationType the annotation type.
     * @param x1 the x coordinate of the top left corner.
     * @param y1 the y coordinate of the top left corner.
     * @param x2 the x coordinate of the bottom right corner.
     * @param y2 the y coordinate of the bottom right corner.
     * @param vertices the polygon annotation vertices.
     * @param uri the URI string.
     * @param key the destination name.
     */
    internal Annotation(
            String annotationType,
            float x1,
            float y1,
            float x2,
            float y2,
            float[] vertices,
            float[] fillColor,
            float transparency,
            String title,
            String contents,
            String uri,
            String key,
            String language,
            String actualText,
            String altDescription) {
        this.annotationType = annotationType;
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;
        this.vertices = vertices;
        this.fillColor = fillColor;
        this.transparency = transparency;
        this.title = title;
        this.contents = contents;
        this.uri = uri;
        this.key = key;
        this.language = language;
        this.actualText = (actualText == null) ? uri : actualText;
        this.altDescription = (altDescription == null) ? uri : altDescription;
    }
}
}   // End of namespace PDFjet.NET
